Stacks: Truncate resource names and labels for host-aware environments

displague opened the pull request:
<!--
Thank you for helping to improve Crossplane!

We strongly recommend you look through our contributor guide at https://git.io/fj2m9
if this is your first time opening a Crossplane pull request. You can find us in
https://slack.crossplane.io/messages/dev if you need any help contributing.
-->

### Description of your changes

Addresses #1337

There may be other scenarios requiring truncation, but this PR addresses the known problem encountered when creating a tenant `StackInstall` resource with a long name in a long namespace that will be processed by a host.

Resource names in the host environment may not exceed 253 characters. Since these names are based on tenant namespace+name they could easily exceed this limitation.
    
There are also some resources created on the host which invoke admission controller behaviors adding resource labels that match resource names. Because labels values are limited to 63 characters the resource names created on the host should not exceed 63 characters.
    
Because the names have been truncated, annotations have been added on the host to help identify the related tenant resource when debugging.
    
This PR adds a truncation package with helpers for specific use-cases. Truncation is performed in a predictable way to make it possible to find resources by label knowing only the original resource names.  The suffix of the truncation is a subset of a lowercased base32 checksum of a SHA1 digest of the original string.  SHA1's security limitations are not a factor here because this is not being used for security purposes, it is chosen for quick and even distribution in the checksum.
    
The suffix is lowercase because resource names must be lowercase. Base32 was chosen to avoid "/" and a conflicting set of alphabetic characters (upper and lower case are used in base64).

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

Fixes #
-->

### How has this code been tested?
<!--
Before reviewers can be confident in the correctness of a pull request,
it needs to tested and shown to be correct. In this section, briefly
describe the testing that has already been done or which is planned.
-->
```sh
cluster/local/hosted_test.sh
export KUBECONFIG=~/.kube/kind-config-kind

# 63 characters
ns="r23456789012345678901234567890123456789012345678901234567890123"

# 63.63.63 characters
resource="s23456789012345678901234567890123456789012345678901234567890123.t23456789012345678901234567890123456789012345678901234567890123.u23456789012345678901234567890123456789012345678901234567890123.u23456789012345678901234567890123456789012345678901234567890"

kubectl create namespace $ns
kubectl crossplane stack install --namespace $ns --cluster 'crossplane/stack-gcp:master' $resource

kind delete cluster --name local-hosted-test
```

### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation] and [examples].
- [x] Reported all new error conditions into the log or as an event, as
  appropriate.

For more about what we believe makes a pull request complete, see our
[definition of done].

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[definition of done]: https://github.com/crossplane/crossplane/tree/master/design/one-pager-definition-of-done.md

jbw976 commented on pkg/controller/stacks/stack/stack.go:
```diff
@@ -722,6 +726,9 @@ func (h *stackHandler) prepareHostAwareJob(j *batch.Job, tokenSecret string) err
 	j.Name = o.Name
 	j.Namespace = o.Namespace
 
+	a := hosted.ObjectReferenceAnnotationsOnHost("stack", h.ext.GetName(), h.ext.GetNamespace())
```
i think the function this is inside of, `prepareHostAwareJob` is a missed remnant of the old job support we used to have and we recently removed.  I think this entire function can be removed.  It's not called from anywhere in the codebase except for tests.

displague replied:
This brings me great joy.

jbw976 commented on pkg/controller/stacks/install/installjob.go:
```diff
@@ -70,15 +70,17 @@ type prepareInstallJobParams struct {
-			Name:      p.name,
-			Namespace: p.namespace,
-			Labels:    p.labels,
+			Name:        p.name,
```
the name that will be used on this job also needs to be truncated to label (<=63) length too I think, if I'm reading this error message correctly:

```
message: 'Job.batch "crossplane-system.crossplane-stack-minimal-gcpv0.2.2-14-g96f534a"
        is invalid: spec.template.labels: Invalid value: "crossplane-system.crossplane-stack-minimal-gcpv0.2.2-14-g96f534a":
        must be no more than 63 characters'
```
I'm interpreting that to mean that the name given to the job, `crossplane-system.crossplane-stack-minimal-gcpv0.2.2-14-g96f534a` is being automatically applied as a label in the job's `spec.template.labels`, so it needs to be truncated.

Here's a snippet of the job and its relevant `spec.template.labels` that seems to confirm that the job's name gets inserted as a `job-name` label inside it's `spec.template.labels`:
```
spec:
  template:
    metadata:
      creationTimestamp: null
      labels:
        controller-uid: 643d1d2f-d6c9-4738-bd86-e8a478268146
        job-name: crossplane-system.crossplane-provider-gcpv0.6.0-rc.29.g762643b
```

The full job yaml output for your convenience can be seen in this gist: https://gist.github.com/jbw976/2c9def142b4e24fd1e39b9f3f439856e

displague replied:
`prepareInstallJob` is called with the name already truncated:
https://github.com/crossplane/crossplane/pull/1338/files#diff-a2cd537f6531321fbef9a92d96b2c425R351-R367

The truncation happens within `ObjectReferenceOnHost`:

https://github.com/crossplane/crossplane/pull/1338/files#diff-33a4f973d281ef627a1c398d9941fc68R73-R84

jbw976 replied:
Ahh, now I see where/how that happens, thanks for pointing that out!

jbw976 commented on pkg/stacks/truncate/truncate.go:
```diff
@@ -0,0 +1,105 @@
+package truncate
+
+import (
+	"crypto/sha1" // nolint:blacklist
```
add a comment for why we're skipping the blacklist lint check here?

displague replied:
Will do. It is also worth pointing out that I think our golangci-lint config may not include the blacklist linter, so this `nolint` is for other environments where this linter is included.

I've noticed warnings from `make` that this linter is not known.

displague replied:
`WARN [runner/nolint] Found unknown linters in //nolint directives: blacklist `

jbw976 replied:
not sure if i fully follow this, if our `make` shows a warning for this linter, and our golangci-lint config doesn't use this linter, should we just remove it? that makes more sense to me, instead of living with a new warning in our default environment.

displague replied:
I tried removing that `nolint` directive but then `gosec` fired.  I've replaced `blacklist` with `gosec` here. `make` has been appeased.

jbw976 commented on pkg/stacks/truncate/truncate_test.go:
```diff
@@ -0,0 +1,200 @@
+			name: "truncate",
+			args: args{str: str253},
+			want: strings.Repeat("1234567890", 5) + "1234567-ij43w",
+		},
```
nit (not blocking): it could be nice to add another sanity (and easy to read) test case for one of the real life scenarios we saw that we are hoping to fix with this new truncation behavior.  These cases test the major scenarios we care about, but they may be a bit hard to read since the last case here has just a little bit of "magic" in them.

displague replied:
I'll add a test case we can refer to as an example.

The testing example in the PR description is similarly obtuse.  The digits make it easy to visually count the character length but don't reasonably convey how this affects "my-really-really-long-namespace.stack-name-is-also-long-version-one-point-oh"

jbw976 commented on pkg/stacks/truncate/truncate_test.go:
```diff
@@ -0,0 +1,200 @@
+	str253 = strings.Repeat("1234567890", 25) + "123"
+)
+
+func Test_truncate(t *testing.T) {
```
great coverage for a lot of possibilities these truncation functions will need to deal with :)

displague commented on pkg/controller/stacks/install/stackinstall.go:
```diff
@@ -344,6 +346,8 @@ func (h *stackInstallHandler) createInstallJob() *batchv1.Job {
 	namespace := i.GetNamespace()
 
 	if hCfg != nil {
+		annotations = hosted.ObjectReferenceAnnotationsOnHost("stackinstall", name, namespace)
```
This was intended as a debugging annotation only.

In the world of edge cases, you got me.

Both a stackinstall and a clusterstackinstall could reside in the same namespace with the same name. I'll have to track "clusterstackinstall" with a different label name.

displague commented:
@jbw976 I believe I've addressed all of your feedback. The added troubleshooting tips are here: https://github.com/crossplane/crossplane/pull/1338/commits/f77f0d5411f324043bea2ef4a017002009a8b0b4

What I added is somewhat generic.  I think we can add more detail to that as cases come up and we find useful patterns.

jbw976 commented on docs/troubleshoot.md:
```diff
@@ -131,3 +132,16 @@ For example, for a Workload object (`workloads.compute.crossplane.io`) named
+
+* [Relationship Labels](https://github.com/crossplane/crossplane/blob/master/design/one-pager-stack-relationship-labels.md)
+* [Owner References](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents)
+* Annotations: `tenant.crossplane.io/{singular}-name` and `tenant.crossplane.io/{singular}-namespace` (_singular_ may be `stackinstall`, `clusterstackinstall` or `stack`)
```
Could we add just a quick (real) concrete example here so people will have more of an understanding of what this is talking about?

jbw976 commented on docs/troubleshoot.md:
```diff
@@ -131,3 +132,16 @@ For example, for a Workload object (`workloads.compute.crossplane.io`) named
+
+Stack resources (including the Stack, service accounts, deployments, and jobs) are usually easy to identify by name. These resource names are based on the name used in the StackInstall or Stack resource.
+
+In some cases, the full name of a Stack resource, which could be up to 253 characters long, can not be represented in the created resources. For example, jobs and deployment names may not exceed 63 characters because these names are turned into resource label values which impose a 63 character limit. Stack created resources whose names would otherwise not be permitted in the Kubernetes API will b ...
```
want to wrap these lines? the rest of the doc appears to be wrapped.

suskin commented on pkg/controller/stacks/stack/stack.go:
```diff
@@ -707,20 +708,8 @@ func (h *stackHandler) prepareHostAwareDeployment(d *apps.Deployment, tokenSecre
 
-	return nil
-}
-func (h *stackHandler) prepareHostAwareJob(j *batch.Job, tokenSecret string) error {
```
Was this just not being used anywhere? The only other references I'm seeing are in tests.

displague replied:
In a resolved comment @jbw976 pointed out that this function was in support of `job` based Stack controllers, which we removed support for in a past release.

suskin commented on pkg/controller/stacks/stack/stack.go:
```diff
@@ -734,7 +723,9 @@ func (h *stackHandler) prepareDeployment(d *apps.Deployment) {
 	// force the deployment to use stack opinionated names and service account
-	name := h.ext.Name + "-controller"
+	suffix := "-controller"
+	name, _ := truncate.Truncate(h.ext.Name, truncate.LabelValueLength-len(suffix), truncate.DefaultSuffixLength)
```
Nitpick: I would normally have expected `truncate.LabelValueLength-len(suffix)` to be more like `truncate.LabelValueLength - len(suffix)` so the operator is easier to see, but maybe this is a go idiom I'm not familiar with yet.

displague replied:
VS code is collapsing the spaces here. :man_shrugging: 

I'll split this into two lines. VS code is not changing the spacing when I do that.

suskin commented on pkg/controller/stacks/stack/stack_test.go:
```diff
@@ -1339,6 +1356,107 @@ func TestProcessDeployment(t *testing.T) {
+				withTokenSecret(corev1.ObjectReference{
+					Name:      resourceName + dashAlphabet,
+					Namespace: namespace + dashAlphabet,
+				})), saSecret(resourceName+dashAlphabet, namespace+dashAlphabet)},
```
Nitpick: noticing the pattern again of having no spaces around the operator

displague replied:
VSCode is collapsing these spaces. I'm going to let this stand since I can not easily split these into separate declarations in the test cases.  If I make each combination into a new constant the names would be even longer: `resourceNamePlusDashAlphabet` :stuck_out_tongue:

displague replied:
`go fmt` seems to be doing this when there are multiple arguments to a function:
https://play.golang.org/p/xVTwhIh6eAb

jbw976 replied:
VS Code collapsing the spaces implies to me that that is the idiomatic thing to do :)

suskin commented on pkg/stacks/relationship.go:
```diff
@@ -33,9 +36,19 @@ const (
-	LabelMultiParentFormat   = LabelMultiParentNSFormat + "-%s"
+
+	// LabelMultiParentFormat defines the format for combining a
+	// LabelMultiParentNSFormat with a named resource
```
Which input is mapped to which placeholder?

suskin commented on pkg/stacks/relationship.go:
```diff
@@ -33,9 +36,19 @@ const (
+	// preserveNSLength is the number of characters using the label name that
+	// will be dedicated to identifying the namespace. This length will include
+	// truncation characters if the namespace exceeds this length.
+	// example: parent.stacks.crossplane.io/{up to 32 chars of NS}-{Name}
```
Excellent detailed comment with example. This helps me reason about the intended behavior of the implementation. :squid:

suskin commented on pkg/stacks/relationship.go:
```diff
@@ -48,16 +61,54 @@ type KindlyIdentifier interface {
 }
 
+// MultiParentLabelPrefix returns the NS specific prefix of a multi-parent label
+// for resources co-owned by a set of Stacks.
```
Originally when I read this, I interpreted it as saying that it would parse a label and return its namespace prefix. Maybe an example of an input and output would be helpful in clearly explaining the reason this method exists.

suskin commented on pkg/stacks/relationship.go:
```diff
@@ -48,16 +61,54 @@ type KindlyIdentifier interface {
+func MultiParentLabelPrefix(stackParent metav1.Object) string {
+	ns := stackParent.GetNamespace()
+
+	// guaranteed not to error based on the lengths of values we are supplying
```
What if a future maintainer changes the lengths?

displague replied:
I've left a warning on preserveNSLength :+1:

suskin commented on pkg/stacks/relationship.go:
```diff
@@ -48,16 +61,54 @@ type KindlyIdentifier interface {
+func MultiParentLabel(stackParent metav1.Object) string {
+	prefix := MultiParentLabelPrefix(stackParent)
+
+	// guaranteed 2 parts based on LabelMultiParentNSFormat
```
Are we using that format here? Since we accept input after it's been formatted, and since this is a public method, I'm not sure we can assume that in the long run. But it may be okay to make the assumption today and revisit it later.

displague replied:
The `MultiParentLabelPrefix` function relies on the `LabelMultiParentPrefix` constant which includes a `/`.  Namespaces can not contain `/` in their resource names. 

If that ever changes, without the rules of label prefixes changing (they allow a single `/`) we would have problems. We have to agree that the world is flat or we have nothing to stand on.

suskin commented on pkg/stacks/relationship_test.go:
```diff
@@ -0,0 +1,310 @@
+	namespace             = "cool-namespace"
+	uidString             = "definitely-a-uuid"
+	resourceName          = "cool-resource"
+	thirtyTwoAs           = strings.Repeat("a", 32)
```
Nitpick: originally I read it as the word `as`

displague replied:
I could come up with another character that can't be confused as a word or abbreviation in this context, but I would also have to update all of the suffixes.  I could add `Letter` to the variable name, but it becomes lengthy.

I think looking at the assignment disambiguates the meaning of the name of this limited use testing variable.

suskin commented on pkg/stacks/relationship_test.go:
```diff
@@ -0,0 +1,310 @@
+		},
+		{
+			name: "Truncated",
+			i:    resource(sixtyThreeAs+"z", sixtyThreeAs+"z", uidString),
```
I  see - because the limit is `63`

suskin commented on pkg/stacks/relationship_test.go:
```diff
@@ -0,0 +1,310 @@
+		{
+			name:        "TruncatedNS",
+			stackParent: resource(sixtyThreeAs+"z", resourceName, uidString),
+			// notice ukl4i is also the suffix with 64 a's because the sum is
```
I didn't follow that - we can get the same suffix for different strings?

displague replied:
I'll revisit this comment.  The intent was to express the following:

If the base32 sum of a digest of  "aaaaaaaaaaa" is "foo" (with a suffix length of 4).

Truncating this string to a length of 11 would result in "aaaaaaaaaaa"
Truncating this string to a length of 8 would result in "aaaa-foo"
Truncating this string to a length of 7 would result in "aaa-foo"
Truncating this string to a length of 5 would result in "a-foo"
 
"aaaaaaaaaaa"

suskin commented on pkg/stacks/truncate/truncate.go:
```diff
@@ -0,0 +1,106 @@
+// An error will be returned if the truncation length is less than the
+// suffixLength, or truncation length is greater than sha1Length, or the suffix
+// length is less than 2.
+func Truncate(str string, length, suffixLength int) (string, error) {
```
I have a really dumb question. Did we already consider using or copying Kubernetes's implementation of this same behavior (that they use for pods, for example)? It seems like maybe the implementation is not long enough to be worth it.

displague replied:
Like Kubernetes, we are using labels to identify the related resources (matchLabels in a deployment refer to the labels in the pod spec).

Our resource names are extra long because we are taking two 253 characters strings (tenant namespace and tenant resource name) and appending them into at most a 253 character slot (the host resource name).  Compounding this further, we need to track these potentially 506 character names as label values, which can be no longer than 63 characters. 

Kubernetes does not offer a solution for this. If you try to create a deployment with a name > 63 characters, the admission controller will reject it.  To apply that same resolution here, if the namespace and name of the tenant Stack resource exceeded 62 characters (losing one for a separator), it would need to be rejected.

This PR seeks to fix the current implementation and avoid changing the existing behavior. Since resources with long name concatenations are not currently supported, this PR should not affect existing resources with the new truncation behavior.  A broader change to how resource names are stored on the host could result in orphaned resources in existing host environments.

Annotations offer sufficient length to store the tenant resource names on the host resource without modification.  The tenant resource names are being stored on the host for debug purposes only.

displague replied:
You may have been referring to the Kubernetes suffix pattern that uses random characters as a suffix (deployment `foo`, replica scale 2 resulting in pod `foo-abc` and `foo-zyx`). 

In order to provide host resource names that could be determined given tenant resource names, randomization was avoided.

Could tenant resource labels have been used to to fetch host resources? Perhaps, but existing resources are not labeled.
Are host resource labels (referring to the tenant) sufficient? Perhaps, but existing resources are not labeled.
Are host resource annotations sufficient? They are not indexed, so a query to find a matching deployment on a host, given a tenant Stack name would need to query all stacks (from all namespaces in a tenant cluster).  Again, these annotations are not present on existing resources.

suskin commented on pkg/stacks/truncate/truncate.go:
```diff
@@ -0,0 +1,106 @@
+	retained := length - suffixLength
+
+	// base64 includes "/" which can only occur once in label names
+	b32 := base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(checksum[:])
```
There are multiple variants to base64, including a [URL-safe variant](https://tools.ietf.org/html/rfc4648#section-5) which does not use `/`. It looks like it's [supported in go](https://golang.org/pkg/encoding/base64/#pkg-variables) too. I would recommend using that instead, unless there's an additional reason to use base32, since base64 is more common.

displague replied:
While label values can include mixed case, resource names can not contain upper-case characters, so a lower-cased Base32 is `[a-z0-7]` fits the job nicely.  I considered mixing the use based on the job, but often the resource name becomes a label value.

I thought one character pattern would be sufficient here and avoid complexity.

suskin commented on pkg/stacks/truncate/truncate_test.go:
```diff
@@ -0,0 +1,211 @@
+		},
+		{
+			name:    "SuffixExceedsSum",
+			args:    args{str: strings.Repeat("12345678890", 21), length: 200, suffixLength: 100},
```
This is a dumb question, but did we consider just padding out to the desired suffix length? : )

displague replied:
I didn't consider that. For our purposes the truncation suffix is locked in.  I added this test and the error response it invokes for logical completeness.

suskin commented on pkg/stacks/truncate/truncate.go:
```diff
@@ -0,0 +1,106 @@
+// a leading hyphen.
+// An error will be returned if the truncation length is less than the
+// suffixLength, or truncation length is greater than sha1Length, or the suffix
+// length is less than 2.
```
How was this minimum chosen?

displague replied:
Suffix length is an argument to `Truncate`. Each of the `Label*` and `ResourceName` helper functions have a predefined length that I pulled from the ether.  36^5 = 60,466,176.  While this may exceed (or not, I couldn't find a reference for it) the max resources per namespace, that is not the goal here.  The goal is to prevent name collisions on truncated resource names, "{really long namespace name}.stack-ms-print" should not be confused with "{same really long namespace name}.stack-ms-paint" when fetching host resources using a source resource as a reference.

any reference to "sha1" should be replaced with hash, digest, or sum.  I don't want to reflect that implementation choice in the comment.

minimum suffix length of 2 offers "-[a-z0-9]"

displague commented on pkg/controller/stacks/hosted/config.go:
```diff
@@ -59,14 +70,30 @@ func NewConfig(hostControllerNamespace, tenantAPIServiceHost, tenantAPIServicePo
 }
 
+// ObjectReferenceAnnotationsOnHost returns a map for use as annotations on the
+// host to identify the named tenant resource
```
As an addendum to https://github.com/crossplane/crossplane/pull/1338#discussion_r395618270

Annotations offer sufficient length to store the tenant resource names on the host resource without modification.  The tenant resource names are being stored on the host for debug purposes only.

displague replied:
I'll weave some of the PR description and comment text into the code comments.  Thanks for pointing out the lack of clarity and telepathy.

displague commented on pkg/controller/stacks/hosted/config.go:
```diff
@@ -59,14 +70,30 @@ func NewConfig(hostControllerNamespace, tenantAPIServiceHost, tenantAPIServicePo
-// ObjectReferenceOnHost maps object with given name and namespace into single controller namespace on Host Cluster.
+// ObjectReferenceOnHost maps object with given name and namespace into single
+// controller namespace on Host Cluster.
+// The resource name on the host cluster is truncated to label value length
```
Yes. I'll defer a deeper explanation to the other review comments. In short, a 253 character tenant resource name (505 with namespace) needs to be representable in a 63 character host label value.

displague replied:
And that 63 character name should be predictable based on the tenant (source) resource name.

displague replied:
While adding some debug examples, I learned namespaces can only be 63 characters. ~505~ 315.

displague replied:
I discovered some caveats to the name length restrictions while doing some testing in response to @jbw976's request for debugging examples.

Namespaces can not exceed 63 characters. 
Deployments and Jobs, the resources that we intend to create on the host, are limited to 63 characters because their names are used in label values.

Of note for #1354, Secrets (as well as ServiceAccounts) abide by the 253 character limit (with 63 character labels) described in https://kubernetes.io/docs/concepts/overview/working-with-objects/names/

Stack and StackInstall don't seem to abide by this naming policy as I was able to create one with 70+ characters in the name, without dots.  Since Stack and StackInstall resources generate deployments and jobs with matching names, they are limited, in practice to 63 characters, which this PR addresses for both hosted and local stack deployments and jobs.

displague commented on pkg/controller/stacks/hosted/config.go:
```diff
@@ -22,9 +22,20 @@ import (
 
 const (
+	// AnnotationTenantNameFmt with a resource singular applied provides the
+	// annotation key used to identify tenant resources by name on the host side
```
CRDs have multiple names. Kind (Foo), plural (foos), and singular (foo) can differ. I am referring to the singular name here.

The expected uses are:

* `tenant.crossplane.io/stackinstall-name: foo"
* `tenant.crossplane.io/stack-name: foo"

displague commented:
@jbw976 @suskin The latest commit responds to your feedback with documentation and comment changes.

displague commented:
Force pushed a few more comment updates that I missed in the last commit, https://github.com/crossplane/crossplane/compare/e62afa38c168c7403fa1a8b38dc0652928a3a4f1..a6b201aee9de31cbd78c091c24c12703df44cf72

jbw976 commented on docs/troubleshoot.md:
```diff
@@ -131,3 +132,81 @@ For example, for a Workload object (`workloads.compute.crossplane.io`) named
+$ name=stack-with-a-really-long-resource-name-so-long-that-it-will-be-truncated
+$ ns=this-is-just-a-really-long-namespace-name-at-the-character-max
+$ kubectl create ns $ns
+$ kubectl crossplane stack install --namespace $ns crossplane/sample-stack-wordpress:0.1.1 $sn
```
what is `$sn` at the end of this line, should that actually be `$name`?

displague replied:
Force pushed the last commit with this tweak.  I used `$sn` locally and replaced it with `$name` in the doc for readability.
